File: src/jimin/week6/주차 요금 계산.py
fees = [180, 5000, 10, 600]
records = ["05:34 5961 IN", "06:00 0000 IN", "06:34 0000 OUT", "07:59 5961 OUT", "07:59 0148 IN", "18:59 0000 IN", "19:09 0148 OUT", "22:59 5961 IN", "23:00 5961 OUT"]
import math
def solution(fees, records):
    answer = []
    # OUT 없다면 23:59 뺌
    # 누적 시간 기준
    car_data = {}
    times = {}

    for i, record in enumerate(records):
        time, car_number, info = record.split(" ")
        print(time, car_number, info)
        if info == "IN":
            car_data[car_number] = time
        elif info == "OUT":
            total_time = get_time(car_data, car_number, time)
            times[car_number] = times.get(car_number, 0) + total_time

    while car_data.keys():
        car_number = list(car_data.keys())[0]
        times[car_number] = times.get(car_number, 0) + get_time(car_data, car_number, "23:59")

    for key, value in sorted(times.items(), key = lambda x:x[0]):
        if value <= fees[0]:
            answer.append(fees[1])
        else:
            answer.append(fees[1] + math.ceil((value- fees[0]) / fees[2]) * fees[3])
    return answer


def get_time(car_data, car_number, time):
    in_hour, in_minute = car_data[car_number].split(":")
    del (car_data[car_number])
    out_hour, out_minute = time.split(":")
    total_time = 0
    if (in_minute <= out_minute):
        total_time += int(out_minute) - int(in_minute)
        total_time += (int(out_hour) - int(in_hour)) * 60
    else:
        total_time += abs((int(in_minute) - int(out_minute)) - 60)
        total_time += (int(out_hour) - int(in_hour) - 1) * 60
    return total_time


print(solution(fees, records))

File: src/daegun/week6/주차 요금 계산.swift
import Foundation

func duration(_ inOut: [String]) -> Int {
    
    let from = inOut[0].split(separator: ":").map { Int(String($0))! }
    let to = inOut[1].split(separator: ":").map { Int(String($0))! }
    
    return (to[0] - from[0]) * 60 + (to[1] - from[1])
}

func solution(_ fees:[Int], _ records:[String]) -> [Int] {
    
    
    var inOut: [Int: [String]] = [:]    // 번호: [입차시각, 출차시각]
    var timeSum: [Int: Int] = [:]       // 번호: 누적주차시간(분)
    for record in records {
        var input = record.split(separator: " ").map { String($0) }
        
        let number = Int(input[1])!
        let time = input[0]
        
        if var existing = inOut[number] {
            inOut[number] = nil
            existing[1] = time
            
            if var count = timeSum[number] {
                timeSum[number] = count + duration(existing)
            } else {
                timeSum[number] = duration(existing)
            }
            
        } else {
            inOut[number] = [time, "23:59"]
        }
    }
    for time in inOut {
        if var count = timeSum[time.key] {
            timeSum[time.key] = count + duration(time.value)
        } else {
            timeSum[time.key] = duration(time.value)
        }
    }
    
    var prices: [Int:Int] = [:]
    for info in timeSum {
        
        var price = fees[1]
        if info.value > fees[0] {
            price += Int(ceil(Float(info.value-fees[0])/Float(fees[2]))) * fees[3]
        }
        
        prices[info.key] = price
    }
    
    return prices.sorted { $0.key < $1.key }.map { $0.value }
}